src/tasks.jl

```diff
@@ -0,0 +1,233 @@
+    n_consume::Int
+
+
+end
+
+## Empty initilaizaiton
+ParticleContainer{Tvi,TInfo}() where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(Dict{String,Function}(),0)
+ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}) where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(manipulators,0)
+#Some initilaizaiton...
+function ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}, n::Int) where {Tvi, TInfo <: AbstractTaskInfo}
+    if !("copy" in keys(manipulators)) manipulators["copy"] = deepcopy end
+    if !("set_retained_vns_del_by_spl!" in keys(manipulators)) manipulators["set_retained_vns_del_by_spl!"] = (x) -> () end
+    ParticleContainer{Tvi,TInfo}( Vector{Trace{Tvi,TInfo}}(undef, n), manipulators, Float64[],Float64[], 0.0, 0)
+
+end
+
+
+Base.collect(pc :: ParticleContainer) = pc.vals # prev: Dict, now: Array
+Base.length(pc :: ParticleContainer)  = length(pc.vals)
+Base.similar(pc :: ParticleContainer{T}) where T = ParticleContainer{T}(0)
+# pc[i] returns the i'th particle
+Base.getindex(pc :: ParticleContainer, i :: Real) = pc.vals[i]
+
+
+
+# registers a new x-particle in the container
+function Base.push!(pc::ParticleContainer, p::Particle{Tvi,TInfo}) where {Tvi,TInfo <: AbstractTaskInfo}
+    push!(pc.vals, p)
+    push!(pc.logWs, 0.0)
+    push!(pc.logpseq,0.0)
+    pc
+end
+
+function extend!(pc::ParticleContainer, n::Int, varInfo, tasks::Task, taskInfo::T) where T <: AbstractTaskInfo
+    # compute total number of particles number of particles
+    n0 = length(pc)
+    ntotal = n0 + n
+
+    # add additional particles and weights
+    vals = pc.vals
+    logWs = pc.logWs
+    logpseq = pc.logpseq
+    resize!(vals, ntotal)
+    resize!(logWs, ntotal)
+    resize!(logpseq, ntotal)
+
+    @inbounds for i in (n0 + 1):ntotal
+        vals[i] = Trace(varInfo,tasks, taskInfo, pc.manipulators["copy"])
+        logWs[i] = 0.0
+        logpseq[i] = 0.0
+    end
+
+    pc
+end
+# clears the container but keep params, logweight etc.
+function Base.empty!(pc::ParticleContainer)
+    pc.vals  = eltype(pc.vals)[]
+    pc.logWs = Float64[]
+    pc.logpseq = Float64[]
+    pc
+end
+
+# clones a theta-particle
+function Base.copy(pc::ParticleContainer)
+    # fork particles
+    vals = eltype(pc.vals)[fork(p, pc.mainpulators["copy"]) for p in pc.vals]
+
+    # copy weights
+    logWs = copy(pc.logWs)
+    logpseq = copy(pc.logpseq)
+
+    ParticleContainer( vals, pc.manipulators, logWs, logpseq, pc.logE, pc.n_consume)
+end
+
+# run particle filter for one step, return incremental likelihood
+function Libtask.consume(pc :: ParticleContainer)
+    # normalisation factor: 1/N
+    z1 = logZ(pc)
+    n = length(pc)
+
+    particles = collect(pc)
+    num_done = 0
+    for i=1:n
+        p = particles[i]
+        score = Libtask.consume(p)
+
+        if in(:logpseq,fieldnames(typeof(p.taskinfo))) && isa(p.taskinfo.logpseq,Float64)
+            set_logpseg(pc,i,p.taskinfo.logpseq)
```

Comment:
I guess this should be
```suggestion
            set_logpseq!(pc, i, p.taskinfo.logpseq)
```